subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/GradleProfilerReporter.java

```diff
@@ -17,57 +17,96 @@
+import java.util.List;
+import java.util.function.Consumer;
 
 public class GradleProfilerReporter implements DataReporter<PerformanceTestResult> {
     private static final String DEBUG_ARTIFACTS_DIRECTORY_PROPERTY_NAME = "org.gradle.performance.debugArtifactsDirectory";
-    private BenchmarkResultCollector resultCollector;
+    private final DelegatingReportGenerator delegatingReportGenerator;
+    private final BenchmarkResultCollector resultCollector;
     private final File debugArtifactsDirectory;
 
     public GradleProfilerReporter(File fallbackDirectory) {
         String debugArtifactsDirectoryPath = System.getProperty(DEBUG_ARTIFACTS_DIRECTORY_PROPERTY_NAME);
         this.debugArtifactsDirectory = Strings.isNullOrEmpty(debugArtifactsDirectoryPath)
             ? fallbackDirectory
             : new File(debugArtifactsDirectoryPath);
-
+        this.delegatingReportGenerator = new DelegatingReportGenerator();
+        this.resultCollector = new BenchmarkResultCollector(delegatingReportGenerator);
     }
 
     @Override
     public void report(PerformanceTestResult results) {
+        PerformanceExperiment experiment = results.getPerformanceExperiment();
+        File baseDir = new File(debugArtifactsDirectory, experiment.getScenario().getTestName().replaceAll("[^a-zA-Z0-9]", "_"));
+        baseDir.mkdirs();
+        delegatingReportGenerator.setDelegates(ImmutableList.of(
+            new CsvGenerator(new File(baseDir, "benchmark.csv"), CsvGenerator.Format.LONG),
+            new HtmlGenerator(new File(baseDir, "benchmark.html"))
+        ));
+
         resultCollector.summarizeResults(line ->
             System.out.println("  " + line)
         );
         try {
-            // TODO Properly pass in the invocation settings for this benchmark here
             InvocationSettings settings = new InvocationSettings.InvocationSettingsBuilder()
-                .setBenchmarkTitle(results.getTestId())
+                .setBenchmarkTitle(experiment.getDisplayName() + " | " + experiment.getScenario().getSimpleClassName())
                 .build();
             resultCollector.write(settings);
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
-    public BenchmarkResultCollector getResultCollector(String displayName) {
-        if (resultCollector == null) {
-            File baseDir = new File(debugArtifactsDirectory, displayName.replaceAll("[^a-zA-Z0-9]", "_"));
-            baseDir.mkdirs();
-            this.resultCollector = new BenchmarkResultCollector(
-                new CsvGenerator(new File(baseDir, "benchmark.csv"), CsvGenerator.Format.LONG),
-                new HtmlGenerator(new File(baseDir, "benchmark.html"))
-            );
-        }
+    public BenchmarkResultCollector getResultCollector() {
         return resultCollector;
     }
 
     @Override
     public void close() {
     }
+
+    private static class DelegatingReportGenerator extends AbstractGenerator {
```

Comment:
Shouldn't this be called a `CompositeGenerator`?

Reply:
Renamed.